File: Programmers/디펜스게임.py
import heapq
"""
라운드가 진행되면서, 무적권을 써야하는 순간마다 i-1라운드까지 중 최대값으로 대체해서 써야한다.
이때 따져줘야하는건
1.무적권이 있는지
2. heapq에 원소가 있는지
3. 있다면, heappop < e인지
를 따져줘서, return해주면 된다.
"""
def solution(n, k, enemy):
    answer = 0
    
    if n>=sum(enemy) or k>=len(enemy):
        return len(enemy)
    
    max_e = []
    for e in enemy:
        flag=True
        if n<e:
            if k>0:
                if max_e:
                    max_num = heapq.heappop(max_e)
                    if max_num[1]<e:
                        heapq.heappush(max_e,(-max_num[1],max_num[1]))
                        n+=e
                        k-=1
                        flag=False
                    else:
                        n+=max_num[1]
                        k-=1
                else:
                    n+=e
                    k-=1
                    flag=False
            else:
                break
        if flag:
            answer+=1
            n-=e
            heapq.heappush(max_e,(-e,e))
        else:
            answer+=1
            n-=e
        
    
    return answer

File: Programmers/디펜스게임.java
/*
enemy를 순회하면서, n-enemy[i]를 계속 진행했을 떄, n<0이 되는 순간 이전의 i+1를 return.
근데, 그 중 k번을 continue할 수 있다.
이때, 가장 많이 진행할 수 있는 경우를 return.

heapq를 써야한다.(enemy의 길이가 100만이므로, 2중 반복문 안됨.)
-> java에선 PriorityQueue.

설명의 마지막 전제인, 모든 라운드를 막을 수 있는 경우(k>=enemy.length) -> return enemy.length;를 예외처리 먼저 해주고,

start round(answer) = k부터 시작한다.
k개의 무적권을 다 쓴 상태 (enemy[k-1]까지 pq에 넣는다.)로 만들고,
k부터 끝까지 돌면서, pq를 pop시키고, 이 값이 enemy[i]와 비교해서
크면 다시 넣은다음 -=enemy[i]를
작으면 enemy[i]를 add하고 -=temp한다.
이렇게 연산을 하고 난 n의 값이 <0이 되면, break해서 answer를 return하면 됨.
*/
import java.util.*;
class Solution {
    public int solution(int n, int k, int[] enemy) {
        if (k>=enemy.length) return enemy.length;
        int answer = k;
        PriorityQueue<Integer> pq = new PriorityQueue<>();
        
        for (int i=0;i<k;i++){
            pq.add(enemy[i]);
        }
        
        for (int i=k;i<enemy.length;i++){
            int temp = pq.remove();
            if (temp>enemy[i]){
                pq.add(temp);
                n-=enemy[i];
            }
            else{
                pq.add(enemy[i]);
                n-=temp;
            }
            if (n<0) break;
            answer+=1;
        }
        
        return answer;
    }
}
